### ISSUE TITLE
Panic in CustomErrorEndpointHandler

### ISSUE THREAD

**Issue by breml** [devopsfaith/krakend#318]
**Describe the bug**

During normal operation of our Krakend based API GW we observed a panic in our version of the CustomErrorEndpointHandler. Our version is based on https://github.com/devopsfaith/krakend/blob/master/router/gin/endpoint.go#L28 and has two changes:
1. we set a value on the context with `c.Set(...)`
2. we do not return the two headers `core.KrakenHeaderName` and `router.CompleteResponseHeaderName`

Other than that, the code is not altered, especially the section that is mentioned in the stack trace (see below) is not changed.

The line of code, that is stated of as the root cause of the panic because of a "invalid memory address or nil pointer dereference" does not make sense, because in [`if isCacheEnabled {`](https://github.com/devopsfaith/krakend/blob/master/router/gin/endpoint.go#L54) the reference variable is a boolean (not a pointer) and therefore it does not seam possible for this variable to be nil.

The comment in the mentioned function `sigpanic` in the [runtime](https://github.com/golang/go/blob/go1.12.12/src/runtime/signal_unix.go#L390) package mentiones, that it "arranges the stack to look like the function where the signal occurred called
 sigpanic" so we asked ourselves, if this "arranging" could be off by one (or even more) lines. This led to the question, if the real cause for the nil pointer is the call to `c.Header(...)` on the next line.

Interesting is the fact, that the request, that is logged (see last line) has a duration of a little bit above of 3 sec and our timeout in the configuration is set to 3 sec.
Based on this observation, it is our guess, that the problem is somehow because of a data race, that only happens, in very rare cases of exact timing, that is if the proxy request just returns before the timeout exceeds. We looked into the code to figure out, where this could lead to a problem (we assumed to reuse of the context objects by gin and the overwriting of the ResponseWriters could be the source of the problem, which would map to the stated assumption above, that the call to `c.Header(...)` could be part of the problem), but we have not been successful so far. 

Have you heard or observed similar panics?
What do you think about the suggestion below in "Additional context"?

**To Reproduce**

We are currently not able to reproduce the error, it only happens very rarely.

**Expected behavior**

The code should not panic.

**Logs**

```
2020/01/14 09:49:14 [Recovery] 2020/01/14 - 09:49:14 panic recovered:
runtime error: invalid memory address or nil pointer dereference
/usr/local/go/src/runtime/panic.go:82 (0x4412c0)
	panicmem: panic(memoryError)
/usr/local/go/src/runtime/signal_unix.go:390 (0x4410ef)
	sigpanic: panicmem()
/usr/src/app/handler_factory.go:64 (0xa4fb5f)
	CustomErrorEndpointHandler.func1: if isCacheEnabled {
/go/pkg/mod/github.com/devopsfaith/krakend-ratelimit@v0.0.0-20190404110207-d63774e96e82/juju/router/gin/gin.go:96 (0x9f7c58)
	NewTokenLimiterMw.func1.1: next(c)
/go/pkg/mod/github.com/gin-gonic/gin@v1.4.0/context.go:124 (0x99c359)
	(*Context).Next: c.handlers[c.index](c)
/usr/src/app/metrics_middleware.go:166 (0xa506cb)
	middleware.func1: c.Next()
/go/pkg/mod/github.com/gin-gonic/gin@v1.4.0/context.go:124 (0x99c359)
	(*Context).Next: c.handlers[c.index](c)
/usr/src/app/router_engine.go:50 (0xa4e72f)
	jsonErrorBody: c.Next()
/go/pkg/mod/github.com/gin-gonic/gin@v1.4.0/context.go:124 (0x99c359)
	(*Context).Next: c.handlers[c.index](c)
/usr/src/app/router_engine.go:46 (0xa4e6d2)
	strictTransportSecurityHeader: c.Next()
/go/pkg/mod/github.com/gin-gonic/gin@v1.4.0/context.go:124 (0x99c359)
	(*Context).Next: c.handlers[c.index](c)
/go/pkg/mod/github.com/gin-gonic/gin@v1.4.0/recovery.go:83 (0x9af629)
	RecoveryWithWriter.func1: c.Next()
/go/pkg/mod/github.com/gin-gonic/gin@v1.4.0/context.go:124 (0x99c359)
	(*Context).Next: c.handlers[c.index](c)
/go/pkg/mod/github.com/gin-gonic/gin@v1.4.0/logger.go:240 (0x9ae6d0)
	LoggerWithConfig.func1: c.Next()
/go/pkg/mod/github.com/gin-gonic/gin@v1.4.0/context.go:124 (0x99c359)
	(*Context).Next: c.handlers[c.index](c)
/go/pkg/mod/github.com/gin-gonic/gin@v1.4.0/gin.go:389 (0x9a5b71)
	(*Engine).handleHTTPRequest: c.Next()
/go/pkg/mod/github.com/gin-gonic/gin@v1.4.0/gin.go:351 (0x9a53a3)
	(*Engine).ServeHTTP: engine.handleHTTPRequest(c)
/usr/local/go/src/net/http/server.go:2774 (0x6c3517)
	serverHandler.ServeHTTP: handler.ServeHTTP(rw, req)
/usr/local/go/src/net/http/server.go:1878 (0x6bf100)
	(*conn).serve: serverHandler{c.server}.ServeHTTP(w, w.req)
/usr/local/go/src/runtime/asm_amd64.s:1337 (0x459060)
	goexit: BYTE	$0x90	// NOP
[GIN] 2020/01/14 - 09:49:14 | 500 |  3.001676485s | xxx.xxx.xxx.xxx | GET      <redacted>
```

**Additional context**

We observed, that in Krakend in most cases, where new timeouts are derived from the global request timeout, only a percentage of the global timeout is used (e.g. [concurrent.go](https://github.com/devopsfaith/krakend/blob/master/proxy/concurrent.go#L16) and [merging.go](https://github.com/devopsfaith/krakend/blob/master/proxy/merging.go#L23)) and we asked ourselves if it would be beneficial to do the same in [endpoint.go](https://github.com/devopsfaith/krakend/blob/master/router/gin/endpoint.go#L35).

**Comment by kpacha** [comment]
interesting... I've never heard about it.

at first glance, I'm not 100% sold on the timeout hypothesis because the handler does wait for the pipe to end before checking the timeout. On the other hand, I think it could be possible to some gorutine to still be running at the time the handler resumes after the pipe execution, so maybe the issue is related to concurrent accesses to the context. it could be fix by passing a context with a lower timeout to the proxy, but then you'll never hit the actual timeout defined by the configuration... If the inner timeout is very close to the actual one, the race conditions could reapear. If it is significantly smaller, the timeout limit defined by the configuration won't ever apply.

since the handler does not read anything from the gin context after calling the proxy pipe, maybe you could replace the line
```
requestCtx, cancel := context.WithTimeout(c, configuration.Timeout)
```

with

```
requestCtx, cancel := context.WithTimeout(c.Copy(), configuration.Timeout)
```

so the pipe context gets derived from a copy of the gin context, making the latter thread safe (at the cost of making an extra copy of every handler context)

I'll make some tests in order to see if I can replicate it using the official binary. In the meantime, it would be great if you check again these files, looking for possible late accesses to the context:

/usr/src/app/metrics_middleware.go
/usr/src/app/router_engine.go
/usr/src/app/router_engine.go

I know the stacktrace says they are just gin middlewares but...

**Comment by breml** [comment]
This problem hit us again so we decided to investigate it more deeply. We have been able to reproduce the problem with an automated test and we finally found the problem. It was in our own modified version of the code for [CustomErrorEndpointHandler](https://github.com/devopsfaith/krakend/blob/master/router/gin/endpoint.go#L28). The thing, that stays "mysterious" is, why the stack trace we got in production did mention the wrong line of code as the source for the nil pointer dereference. The file and the function have been correct, but the line of code is off by 8 lines.
In the tests we performed (using `go test`), we finally got the right position in the source.
